### ISSUE TITLE
Vim9: no error for concatenating dictionary with string inside :def function

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#6655]
**Describe the bug**

In Vim9 script, a dictionary is automatically coerced into a string when concatenated to another string inside a `:def` function.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        def Func()
            let d = {'key': 'val'}
            echo 'aaa ' .. d .. ' bbb'
        enddef
        Func()
    EOF
    )

No error is raised.  This is echo'ed:

    aaa {'key': 'val'} bbb

**Expected behavior**

Nothing is echo'ed.  An error is raised:

    E731: using Dictionary as a String

**Environment**

 - Vim version: 8.2 Included patches: 1-1382
 - OS: Ubuntu 16.04.6 LTS
 - Terminal: XTerm(358)

**Additional context**

At the script-level, an error is correctly raised:
```vim
vim9script
let d = {'key': 'val'}
echo 'aaa ' .. d .. ' bbb'
```
    E731: using Dictionary as a String

---

As well as in a legacy function:
```vim
vim9script
fu Func()
    let d = {'key': 'val'}
    echo 'aaa ' .. d .. ' bbb'
endfu
Func()
```
    E731: using Dictionary as a String

---

Not raising an error in a `:def` function is inconsistent with the 2 other contexts; and if an error is raised later because of this automatic coercion, it's much harder to spot its real origin.

**Comment by brammool** [comment]
> **Describe the bug**
> 
> In Vim9 script, a dictionary is automatically coerced into a string when concatenated to another string inside a `:def` function.
> 
> **To Reproduce**
> 
> Run this shell command:
> 
>     vim -Nu NONE -S <(cat <<'EOF'
>         vim9script
>         def Func()
>             let d = {'key': 'val'}
>             echo 'aaa ' .. d .. ' bbb'
>         enddef
>         Func()
>     EOF
>     )
> 
> No error is raised.  This is echo'ed:
> 
>     aaa {'key': 'val'} bbb
> 
> **Expected behavior**
> 
> Nothing is echo'ed.  An error is raised:
> 
>     E731: using Dictionary as a String

The Vim9 docs actually mention that the arguments of ".." are converted
to a string:

	When using `..` for string concatenation the arguments are always converted to
	string. >
		'hello ' .. 123  == 'hello 123'
		'hello ' .. v:true  == 'hello true'

For numbers and a few other types this is nice, so how about
dictionaries and lists?

Your report suggests you expected an error.  That's because you don't
really expect a composite type to be turned into a string.  I suppose
you can say the same for Job and a Channel.

How about a Funcref?  Perhaps we should also disallow that, otherwise
you don't get an error when forgetting the ():

	let result = 'the ' .. Result
	# should be:
	let result = 'the ' .. Result()


-- 
hundred-and-one symptoms of being an internet addict:
145. You e-mail your boss, informing him you'll be late.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
> The Vim9 docs actually mention that the arguments of ".." are converted to a string:

Ah you are right.  It is documented.

I can see how convenient it is, so maybe it should keep working like that.

I just had difficulties finding the cause of an issue; it would have been easier if Vim had raised an error earlier, right when I tried to wrongly concatenate a dictionary with a string.  Although that doesn't mean much; I still need to learn how to properly debug `:def` functions; `:breakadd` (and `:profile`) don't seem to work, but that's another story.

However, to be consistent, if Vim coerces a dictionary into a string inside a `:def` function, I think it should do the same at the script-level.  This is not the case right now:
```vim
vim9script
let d = {'key': 'val'}
echo 'aaa ' .. d .. ' bbb'
```
    E731: using Dictionary as a String

---

> For numbers and a few other types this is nice, so how about
> dictionaries and lists?
>
> Your report suggests you expected an error.  That's because you don't
> really expect a composite type to be turned into a string.  I suppose
> you can say the same for Job and a Channel.
>
> How about a Funcref?  Perhaps we should also disallow that, otherwise
> you don't get an error when forgetting the ():

I don't know about lists and dictionaries, but I guess we could be more strict about types for which it will be rarely useful, like jobs, channels, and funcrefs...  As you said in another comment, it's easier to be strict at first, then relax the rules if needed later, rather than the opposite (because of backwards compatibility).

**Comment by brammool** [comment]
> > The Vim9 docs actually mention that the arguments of ".." are converted to a string:
> 
> Ah you are right.  It is documented.
> 
> I can see how convenient it is, so maybe it should keep working like that.
> 
> I just had difficulties finding the cause of an issue; it would have
> been easier if Vim had raised an error earlier, right when I tried to
> wrongly concatenate a dictionary with a string.  Although that doesn't
> mean much; I still need to learn how to properly debug `:def`
> functions; `:breakadd` (and `:profile`) don't seem to work, but that's
> another story.
> 
> However, to be consistent, if Vim coerces a dictionary into a string
> inside a `:def` function, I think it should do the same at the
> script-level.  This is not the case right now:
> ```vim
> vim9script
> let d = {'key': 'val'}
> echo 'aaa ' .. d .. ' bbb'
> ```
>     E731: using Dictionary as a String
> 
> ---
> 
> > For numbers and a few other types this is nice, so how about
> > dictionaries and lists?
> >
> > Your report suggests you expected an error.  That's because you don't
> > really expect a composite type to be turned into a string.  I suppose
> > you can say the same for Job and a Channel.
> >
> > How about a Funcref?  Perhaps we should also disallow that, otherwise
> > you don't get an error when forgetting the ():
> 
> I don't know about lists and dictionaries, but I guess we could be
> more strict about types for which it will be rarely useful, like jobs,
> channels, and funcrefs...  As you said in another comment, it's easier
> to be strict at first, then relax the rules if needed later, rather
> than the opposite (because of backwards compatibility).

We could follow Javascript/Typescript, but then you'll get something
like:
	[Object: object]
which is totally useless.

For Job and Channel it's hard to come up with a good string, other then
what's used for debugging.  And most likely it's a mistake.

Since one can always use string() if a string is really desired, the
rule to only do automatic conversion for simple types is probably the
best compromise between easy of use and avoiding mistakes.

-- 
hundred-and-one symptoms of being an internet addict:
147. You finally give up smoking...because it made the monitor dirty.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by brammool** [comment]
Implemented with patch 8.2.1435
